nibabel/affines.py

```diff
@@ -80,8 +83,18 @@ def apply_affine(aff, pts):
     # rzs == rotations, zooms, shears
     rzs = aff[:-1, :-1]
     trans = aff[:-1, -1]
-    res = np.dot(pts, rzs.T) + trans[None, :]
-    return res.reshape(shape)
+
+    if inplace:
+        try:
+            np.dot(pts, rzs.T, out=pts)
+            pts += trans[None, :]
+        except ValueError:
+            # Fallback on the not inplace version.
+            pts = np.dot(pts, rzs.T) + trans[None, :]
+    else:
+        pts = np.dot(pts, rzs.T) + trans[None, :]
```

Comment:
If anyone has a cleaner code for this, I'm a taker.

Reply:
Under what circumstances do you end up with the `ValueError` (that wouldn't raise for the `out=None` case)?

Reply:
According to Numpy's documentation:
> out ndarray, optional
Output argument. This must have the exact kind that would be returned if it was not used. In particular, it must have the right type, must be C-contiguous, and its dtype must be the dtype that would be returned for dot(a,b). This is a performance feature. Therefore, if these conditions are not met, an exception is raised, instead of attempting to be flexible.

https://numpy.org/doc/stable/reference/generated/numpy.dot.html

Reply:
It does happen in the tests. That's why I added it. If we don't want it in the `affines.apply_affine` I can check it in the Tractogram class.

Reply:
A couple things:

1) If `np.dot(pts, rzs.T, out=pts)` succeeds but `pts += trans[None, :]` fails, then the result will be `pts = pts @ rzs.T @ rzs.T + trans[None, :]`. This seems unlikely but we should be a bit defensive.
2) I would probably not duplicate `pts = np.dot(pts, rzs.T) + trans[None, :]` and instead encode the fallback explicitly in the branching structure. This is debatable, aesthetically, so feel free to reject.

Combining the two (and using `@` instead of `np.dot`):

```suggestion
    if inplace:
        try:
            np.dot(pts, rzs.T, out=pts)
        except ValueError:
            inplace = False
        else:
            pts += trans[None, :]
    if not inplace:
        pts = pts @ rzs.T + trans[None, :]
```

Reply:
I like it. Thanks.